models/user_repo_unit.go

```diff
@@ -0,0 +1,1079 @@
+// * user collaboration status
+// * user ownership of the repository
+// * user site admin status
+// * which repository units are enabled (e.g. issues, PRs, etc.)
+
+// UserRepoUnit is an explosion (cartesian product) of all user permissions
+// on all repositories, and all types of relevant unit types.
+type UserRepoUnit struct {
+	UserID               int64      `xorm:"pk"`
+	RepoID               int64      `xorm:"pk INDEX"`
+	Type                 UnitType   `xorm:"pk"`
+	Mode                 AccessMode `xorm:"NOT NULL"`
+}
+
+// UserRepoUnitWork is a table used for temporarily accumulate all the work performed
+// while processing a batch. Ideally, this would be a temporary (no storage) table.
+// Records are grouped by BatchID in order to prevent any kind of collision.
+// Lack of primary key is intentional; this table is not intended for replication
+// as it should never contain rows after the transaction is completed.
+type UserRepoUnitWork struct {
+	BatchID              int64      `xorm:"NOT NULL INDEX"`
+	UserID               int64      `xorm:"NOT NULL"`
+	RepoID               int64      `xorm:"NOT NULL"`
+	Type                 UnitType   `xorm:"NOT NULL"`
+	Mode                 AccessMode `xorm:"NOT NULL"`
+}
+
+// UserRepoUnitBatchNumber provides in a safe way unique ID values
+// for the batch number in case we are in a multi-server environment.
+// It's a 63-bit number, so good luck reaching the maximum value
+// (300 million years at 1000 requests per second, if you want to know).
+// It's a makeshift replacement for an actual database sequence.
+type UserRepoUnitBatchNumber struct {
+	ID int64 `xorm:"pk autoincr"`
+}
+
+var (
+	// This list may be updated as needed. Requisites are:
+	// - New units need to abide by the same rules as the others
+	// - Units from all repos must be rebuilt after this change
+	supportedUnits = []UnitType {
+		UnitTypeCode,
+		UnitTypeIssues,
+		UnitTypePullRequests,
+		UnitTypeWiki,
+	}
+
+	// Pre-built SQL condition to filter-out unsupported unit types
+	// "repo_unit.`type` IN (UnitTypeCode, UnitTypeIssues, etc.)"
+	userRepoUnitSupported		string
+)
+
+func init() {
+	// Build some data that we will use quite often
+	uts := make([]string, len(supportedUnits))
+	for i, su := range supportedUnits {
+		uts[i] = fmt.Sprintf("%d",su)
+	}
+	userRepoUnitSupported = "repo_unit.`type` IN (" + strings.Join(uts, ",") + ")"
+}
+
+// RebuildRepoUnits will rebuild all permissions to a given repository for all users
+func RebuildRepoUnits(e Engine, repo *Repository) error {
+
+	batchID, err := userRepoUnitStartBatch(e)
+	if err != nil {
+		return fmt.Errorf("userRepoUnitStartBatch: %v", err)
+	}
+
+	log.Trace("RebuildUserUnits: rebuilding permissions for repository %d on batch %d", repo.ID, batchID)
+
+	// Make sure we start from scratch; we intend to recreate all pairs
+	_, err = e.Delete(&UserRepoUnit{RepoID: repo.ID})
```

Comment:
This would be a concurrency issue. It should only be done in the end of the batch while copying from the work table in same session?

Reply:
You're right about the concurrency problem. Moving the `DELETE` down until the last moment will certainly help. However, I'm thinking of implementing some locking mechanism as well. Since we're not enforcing `TRANSACTION MODE = SERIALIZED` (or whatever name this mode has in every other database), source information can change in the middle of our calculations (this is true for other processes in Gitea as well). I'm still evaluating options for that (a queue, an auxiliary table, etc.).

Reply:
If you use one db session and commit in the end, would that suffice?

Reply:
Not according to my tests. e.g. in PostgreSQL:

#### Initial state of table `a`

| id  |    value     |
|-----|--------------|
|  1  |   'hello'    |
|  2  |   'bye bye'  |

* Sesion 1: `BEGIN TRANSACTION;`
* Sesion 1: `SELECT value FROM a WHERE ID = 1;` gets `'hello'`.
    * Sesion 2: `BEGIN TRANSACTION;`
    * Sesion 2: `UPDATE a SET value = 'good bye' WHERE id = 2;`
    * Sesion 2: `COMMIT;`
* Sesion 1: `SELECT value FROM a WHERE ID = 2;` gets `'good bye'`, expected `'bye bye'`.

I mean, the problem is that there will potentially be certainly be several sessions updating data on the instance (a rare case, but two admins might be creating teams for the same organization at the same time).